File: src/domain/events.ts
import { ulid } from "ulid";
import type { EventPayloadMap, EventRecord, EventType } from "../types";

/** Shared event factory for constructing EventRecord instances. */
export function makeEvent<T extends EventType>(
  actor: string,
  ts: string,
  type: T,
  taskId: string,
  payload: EventPayloadMap[T],
): EventRecord {
  return {
    event_id: ulid(),
    ts,
    actor,
    type,
    task_id: taskId,
    payload: payload as Record<string, unknown>,
  };
}

File: src-rust/src/domain/events.rs
use crate::types::{EventRecord, EventType};
use serde_json::{Map, Value};
use ulid::Ulid;

pub fn make_event(
    actor: &str,
    ts: &str,
    event_type: EventType,
    task_id: &str,
    payload: Map<String, Value>,
) -> EventRecord {
    let id = Ulid::new().to_string();
    EventRecord {
        id: Some(id.clone()),
        event_id: Some(id),
        ts: ts.to_string(),
        actor: actor.to_string(),
        event_type,
        task_id: task_id.to_string(),
        payload,
    }
}
